src/DataStructures/TempBuffer.hpp

```diff
@@ -39,3 +39,21 @@ template <typename TagList>
 struct TempBuffer<TagList, false> : Variables<TagList> {
   using Variables<TagList>::Variables;
 };
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief A common operation for checking the size and resizing a memory buffer
```

Comment:
I don't like the word `resize` here because this function doesn't behave the way `resize` works for `std::vector` (https://en.cppreference.com/w/cpp/container/vector/resize). I'm trying to think of a better word, and also that this should probably be a member function to be more in line with the STL interface. Maybe `destructive_resize`? I can see some of our classes benefiting from a `resize` function that's non-destructive.

I'd like to have a bit of dialogue on this point before making code changes to the PR to reduce your workload :)

Reply:
I'd be fine with a renaming of the function.

I'm reluctant, though, to move it to a member function (presumably your suggestion is a member function of `VectorImpl`?), as I'd like to later add in overloads that work on `Tensors`.

Reply:
How do you plan on resizing a `Tensor`? It's compile-time size.

Reply:
Sorry, I didn't mean the dimensionality of the `Tensor`, but the size of each of its components.

Reply:
I don't like that. That should be done via a loop since overloading this function means it behaves completely differently on different containers. I find this leads to confusing behavior and makes the code more difficult to use.

Reply:
Perhaps a differently named resizing function, then? Or even both as member functions of their respective classes? I just note that effectively the same ~4 lines are repeated most places a tensor is taken by pointer.

Reply:
Maybe, I'm not so convinced that just using `alg::for_each` isn't the best solution, depending on what you want to do. I'm a bit confused by your comment about the 4 lines anywhere a pointer is passed by `not_null`. An important clarification for me is are you trying to replace code like:
```cpp
  if (UNLIKELY(get_size(get<0, 0, 0>(*phi)) != get_size(get(lapse)))) {
    *phi = tnsr::iaa<DataType, SpatialDim, Frame>(get_size(get(lapse)));
  }
```
I'd be open to a member function of `Tensor<X, ...>` called `assign` (in line with the std::vector interface) that takes an object of `X` and assigns it to all the components. I guess you want the conditional assign, I'm not sure if I really like that wrapped. I'm currently undecided :) Maybe an `assign_if` is a solution. Not sure

Reply:
Yes, it's the pasted code I'd be trying to replace with a utility call. It seems to me that the assign also isn't what's desired here -- the pasted code just allocates without copies, which is what we'd want in many cases.

Reply:
Okay, so you want something that just calls a constructor inside an if statement?

Reply:
Yes. I realize the savings here is minor, but `buffer->resize_function(size);` seems better to me than what we've got in many places.

Reply:
Okay, I'd be happy with a small wrapper function around `destructive_resize` (I think the `unlikely` part can be left out of the name). Maybe it should be in something like `DataStructures/Tensor/Helpers.hpp`?

Reply:
That sounds good. I'll put that in the later PRs in this series that perform the resize on `Tensor`s.

Reply:
Awesome, thank you :D